File: mobile/src/types/community.ts
import type { User } from "./auth";

export type CommunityRole = "OWNER" | "MODERATOR" | "MEMBER";
export type CommunityVisibility = "public" | "private";

export const COMMUNITY_CATEGORIES = [
  "Java Programming",
  "Python Programming",
  "Web Development",
  "Cyber Security",
  "Data Science",
  "Competitive Programming",
  "Placement Preparation",
  "Other",
] as const;

export type CommunityCategory = (typeof COMMUNITY_CATEGORIES)[number];

export interface Community {
  _id: string;
  name: string;
  slug: string;
  description: string;
  bannerImage?: string;
  category: CommunityCategory;
  tags: string[];
  visibility: CommunityVisibility;
  owner: Pick<User, "_id" | "fullName" | "rollNumber" | "profilePicture">;
  moderators: string[];
  memberCount: number;
  extensionPoints: {
    chatEnabled: boolean;
    resourcesEnabled: boolean;
    notificationsEnabled: boolean;
  };
  membershipRole: CommunityRole | null;
  isMember: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CommunityMember {
  _id: string;
  communityId: string;
  userId: Pick<
    User,
    "_id" | "fullName" | "rollNumber" | "department" | "academicYear" | "profilePicture"
  >;
  role: CommunityRole;
  joinedAt: string;
}

export interface PaginatedCommunities {
  items: Community[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

File: backend-java/src/main/java/com/studyconnect/backend/entity/Community.java
package com.studyconnect.backend.entity;

import com.studyconnect.backend.entity.enums.CommunityCategory;
import com.studyconnect.backend.entity.enums.CommunityVisibility;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.FieldType;
import org.springframework.data.mongodb.core.mapping.MongoId;

@Getter
@Setter
@NoArgsConstructor
@Document(collection = "communities")
@CompoundIndex(name = "community_category_visibility_created", def = "{'category': 1, 'visibility': 1, 'createdAt': -1}")
public class Community {

    @Id
    @MongoId(FieldType.OBJECT_ID)
    private String id;

    @Indexed(unique = true)
    private String name;

    @Indexed(unique = true)
    private String slug;

    private String description = "";
    private String bannerImage;

    @Indexed
    private CommunityCategory category;

    private List<String> tags = new ArrayList<>();

    @Indexed
    private CommunityVisibility visibility = CommunityVisibility.PUBLIC;

    @Indexed
    private String owner;

    private List<String> moderators = new ArrayList<>();
    private long memberCount = 0;
    private ExtensionPoints extensionPoints = new ExtensionPoints();

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;

    @Getter
    @Setter
    @NoArgsConstructor
    public static class ExtensionPoints {
        private boolean chatEnabled;
        private boolean resourcesEnabled;
        private boolean notificationsEnabled;
    }
}
